### PR TITLE
Edgesite: cluster on edge implement

### PR DESCRIPTION
**Pull request by liufen90**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Implement PR #2650 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: build/edgesite/proxy-agent.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    command: [ "/proxy-agent"]
+    args: [
+      "--ca-cert=/root/ca/rootCA.crt",
+      "--agent-cert=/root/certs/proxy-agent.crt",
+      "--agent-key=/root/certs/proxy-agent.key",
+      "--proxy-server-host=${PROXY_SERVER_IP}",
+      "--agent-identifiers=host=${KUBE_APISERVER_IP}",
+      ]
+    volumeMounts:
+    # 容器内路径
```
Comment (fisherxu) [inline build/edgesite/proxy-agent.yaml:21]: please have a look

File: build/edgesite/proxy-server.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      containerPort: 8091
+      hostPort: 8091
+    - name: healthport
+      containerPort: 8092
+      hostPort: 8092
+    - name: adminport
+      containerPort: 8095
+      hostPort: 8095
+    volumeMounts:
+    # 容器内路径
```
Comment (fisherxu) [inline build/edgesite/proxy-server.yaml:37]: ditto

**Comment by liufen90** [comment]
@fisherxu please have a look

File: build/edgesite/edgesite-agent.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+apiVersion: v1
+kind: Pod
```
Comment (fisherxu) [inline build/edgesite/edgesite-agent.yaml:2]: Do we need to use deployment?

File: build/edgesite/edgesite-server.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1
+kind: Pod
```
Comment (fisherxu) [inline build/edgesite/edgesite-server.yaml:2]: ditto

File: build/edgesite/server-build.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Build the proxy-server binary
+FROM golang:1.13.4 as builder
```
Comment (fisherxu) [inline build/edgesite/server-build.Dockerfile:2]: Use a higher version of golang?

File: edgesite/cmd/edgesite-agent/main.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:2]: ?

File: edgesite/hack/go-license-header.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/*
+Copyright The Kubernetes Authors.
```
Comment (fisherxu) [inline edgesite/hack/go-license-header.txt:2]: Please use kubeedge header.

File: edgesite/konnectivity-client/go.mod [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module sigs.k8s.io/apiserver-network-proxy/konnectivity-client
+
+go 1.13
```
Comment (fisherxu) [inline edgesite/konnectivity-client/go.mod:3]: 1.14?

Comment (liufen90) [reply edgesite/konnectivity-client/go.mod]: I have deleted this file

File: edgesite/hack/go-license-header.txt (hunk 7 above)
Comment (liufen90) [reply edgesite/hack/go-license-header.txt]: I have deleted this file

File: build/edgesite/edgesite-agent.yaml (hunk 3 above)
Comment (liufen90) [reply build/edgesite/edgesite-agent.yaml:2]: now we only use one pod, no need to use deployment

File: build/edgesite/edgesite-server.yaml (hunk 4 above)
Comment (liufen90) [reply build/edgesite/edgesite-server.yaml:2]: now we only use one pod, no need to use deployment

File: edgesite/cmd/edgesite-agent/main.go (hunk 6 above)
Comment (liufen90) [reply edgesite/cmd/edgesite-agent/main.go]: done

File: build/edgesite/server-build.Dockerfile (hunk 5 above)
Comment (liufen90) [reply build/edgesite/server-build.Dockerfile]: done

File: Makefile [hunk 9]
```diff
@@ -286,17 +284,13 @@ edgeimage:
...
-	curl -L -o ./build/edgesite/tmp/qemu-${QEMU_ARCH}-static.tar.gz https://github.com/multiarch/qemu-user-static/releases/download/v3.0.0/qemu-${QEMU_ARCH}-static.tar.gz
-	tar -xzf ./build/edgesite/tmp/qemu-${QEMU_ARCH}-static.tar.gz -C ./build/edgesite/tmp
-	docker build -t kubeedge/edgesite:${IMAGE_TAG} \
-	--build-arg GO_LDFLAGS=${GO_LDFLAGS} \
-	--build-arg BUILD_FROM=${ARCH}/golang:1.14-alpine3.11 \
-	--build-arg RUN_FROM=${ARCH}/docker:dind \
-	-f build/edgesite/Dockerfile .
+.PHONY: edgesite-server-image
+edgesite-server-image:
+	docker build . --build-arg ARCH=${ARCH} -f build/edgesite/server-build.Dockerfile -t liufen90/edgesite-server-${ARCH}:${IMAGE_TAG}
```
Comment (fisherxu) [inline Makefile:289]:
```suggestion
	docker build . --build-arg ARCH=${ARCH} -f build/edgesite/server-build.Dockerfile -t kubeedge/edgesite-server-${ARCH}:${IMAGE_TAG}
```

File: Makefile [hunk 10]
```diff
@@ -286,17 +284,13 @@ edgeimage:
...
+
+.PHONY: edgesite-agent-image
+edgesite-agent-image:
+	docker build . --build-arg ARCH=${ARCH} -f build/edgesite/agent-build.Dockerfile -t liufen90/edgesite-agent-${ARCH}:${IMAGE_TAG}
```
Comment (fisherxu) [inline Makefile:293]: ditto

File: build/edgesite/edgesite-agent.yaml (hunk 3 above)
Comment (fisherxu) [reply build/edgesite/edgesite-agent.yaml:2]: Please use deployment, thanks.

File: build/edgesite/edgesite-server.yaml (hunk 4 above)
Comment (fisherxu) [reply build/edgesite/edgesite-server.yaml:2]: Please use deployment, thanks.

File: Makefile (hunk 9 above)
Comment (liufen90) [reply Makefile]: done

File: Makefile (hunk 10 above)
Comment (liufen90) [reply Makefile]: done

File: build/edgesite/edgesite-server.yaml (hunk 4 above)
Comment (liufen90) [reply build/edgesite/edgesite-server.yaml]: done

Comment (liufen90) [reply build/edgesite/edgesite-server.yaml]: done

File: edgesite/cmd/edgesite-agent/main.go [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+		flags.AddGoFlag(fl)
+	})
+	if err := command.Execute(); err != nil {
+		klog.Errorf("error: %v\n", err)
+		klog.Flush()
+		os.Exit(1)
+	}
+}
+
+type GrpcProxyAgentOptions struct {
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:63]: Can't this struct be imported from anp？

File: edgesite/cmd/edgesite-agent/main.go [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+	agentID          string
+	agentIdentifiers string
+	syncInterval     time.Duration
+	probeInterval    time.Duration
+
+	// file contains service account authorization token for enabling proxy-server token based authorization
+	serviceAccountTokenPath string
+}
+
+func (o *GrpcProxyAgentOptions) ClientSetConfig(dialOptions ...grpc.DialOption) *agent.ClientSetConfig {
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:87]: ditto

File: edgesite/cmd/edgesite-agent/main.go [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+		AgentID:                 o.agentID,
+		AgentIdentifiers:        o.agentIdentifiers,
+		SyncInterval:            o.syncInterval,
+		ProbeInterval:           o.probeInterval,
+		DialOptions:             dialOptions,
+		ServiceAccountTokenPath: o.serviceAccountTokenPath,
+	}
+}
+
+func (o *GrpcProxyAgentOptions) Flags() *pflag.FlagSet {
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:99]: ditto

File: edgesite/cmd/edgesite-agent/main.go [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+	flags.IntVar(&o.adminServerPort, "admin-server-port", o.adminServerPort, "The port the admin server is listening on.")
+	flags.StringVar(&o.agentID, "agent-id", o.agentID, "The unique ID of this agent. Default to a generated uuid if not set.")
+	flags.DurationVar(&o.syncInterval, "sync-interval", o.syncInterval, "The initial interval by which the agent periodically checks if it has connections to all instances of the proxy server.")
+	flags.DurationVar(&o.probeInterval, "probe-interval", o.probeInterval, "The interval by which the agent periodically checks if its connections to the proxy server are ready.")
+	flags.StringVar(&o.serviceAccountTokenPath, "service-account-token-path", o.serviceAccountTokenPath, "If non-empty proxy agent uses this token to prove its identity to the proxy server.")
+	flags.StringVar(&o.agentIdentifiers, "agent-identifiers", o.agentIdentifiers, "Identifiers of the agent that will be used by the server when choosing agent. N.B. the list of identifiers must be in URL encoded format. e.g.,host=localhost&host=node1.mydomain.com&cidr=127.0.0.1/16&ipv4=1.2.3.4&ipv4=5.6.7.8&ipv6=:::::")
+	return flags
+}
+
+func (o *GrpcProxyAgentOptions) Print() {
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:117]: ditto

File: edgesite/cmd/edgesite-agent/main.go [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+	klog.V(1).Infof("HealthServerPort set to %d.\n", o.healthServerPort)
+	klog.V(1).Infof("AdminServerPort set to %d.\n", o.adminServerPort)
+	klog.V(1).Infof("AgentID set to %s.\n", o.agentID)
+	klog.V(1).Infof("SyncInterval set to %v.\n", o.syncInterval)
+	klog.V(1).Infof("ProbeInterval set to %v.\n", o.probeInterval)
+	klog.V(1).Infof("ServiceAccountTokenPath set to %q.\n", o.serviceAccountTokenPath)
+	klog.V(1).Infof("AgentIdentifiers set to %s.\n", util.PrettyPrintURL(o.agentIdentifiers))
+}
+
+func (o *GrpcProxyAgentOptions) Validate() error {
```
Comment (fisherxu) [inline edgesite/cmd/edgesite-agent/main.go:133]: ditto

**Review by fisherxu — COMMENTED** [review]
/lgtm

**Comment by fisherxu** [comment]
Open an issue #2869  to track import issue.

**Pull request merged**